### PR TITLE
Created HDF5 based dataset

### PR DESCRIPTION
**Pull request by peastman**
This is a rough draft of the HDF5 based dataset discussed in #214.  It can handle arbitrarily large datasets, even ones that are too large to fit in memory.  Are you interested in developing this into a supported feature?

## Description
I added a HDF5Dataset class that extends AtomicDataset.  It loads data from one or more HDF5 files with the format described at https://github.com/torchmd/torchmd-net/blob/main/torchmdnet/datasets/hdf.py.  The only data fields it currently supports are positions, atom types, energies, and forces.  Others could easily be added.

There's a lot that could be cleaned up.  I've only implemented the features that were needed for my own work.  That's especially true of `statistics()`, which only computes a subset of the available statistics.  It could be extended to compute the rest of them, but in the long term it might be better to rework the main statistics routine to support streaming calculations so it can work with any dataset class.

The class constructs AtomicData objects as needed in `get()`, which adds overhead.  You can partly compensate for that by increasing `dataloader_num_workers` to use multiple processes.

Computing statistics on large datasets can be slow.  I use `dataset_statistics_stride` to reduce the cost.

## Motivation and Context
All of the existing dataset classes require all the data to be held in memory at once.  This limits how much data can be used for training.
Resolves: #214.

## How Has This Been Tested?
I used it to train a model and it generally seems to work, but I haven't done extensive testing or written unit tests yet.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds or improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation improvement (updates to user guides, docstrings, or developer docs)

## Checklist:

- [x] My code follows the code style of this project and has been formatted using `black`.
- [x] All new and existing tests passed, including on GPU (if relevant).
- [x] I have added tests that cover my changes (if relevant).
- [ ] The option documentation (`docs/options`) has been updated with new or changed options.
- [ ] I have updated `CHANGELOG.md`.
- [ ] I have updated the documentation (if relevant).

### CODE REVIEW

**Comment by Linux-cpp-lisp** [comment]
Hi @peastman ,

Thanks very much for posting this! I am away from code development right now and unfortunately will not be able to take a closer look for a little while, but at a glance this looks like a promising start.

> You can partly compensate for that by increasing dataloader_num_workers to use multiple processes.

Did you find this to be true in practice? I've had trouble getting anything but slowdowns from adding in the multiprocessing.

**Comment by peastman** [comment]
Thanks!  There's no hurry.  Whenever you get around to looking at it.

I did find some benefit to using more workers.  My GPU utilization increased from 45% with one worker to 55% with three workers.  By increasing the batch size I was able to increase it further, up to about 85%.

**Comment by peastman** [comment]
Can I bump this issue?  The inability to work with large datasets is really a showstopper for me.

File: nequip/data/_dataset/_hdf5_dataset.py
Comment (Linux-cpp-lisp) [inline nequip/data/_dataset/_hdf5_dataset.py:26]: -> `num_frames`

File: nequip/data/_dataset/_hdf5_dataset.py
Comment (Linux-cpp-lisp) [inline nequip/data/_dataset/_hdf5_dataset.py:126]: See https://github.com/mir-group/pytorch_runstats / https://pytorch-runstats.readthedocs.io/en/latest/index.html  

which can do this for `rms` and `mean` already with unit tests and correct numerics.  We already depend on this library.

Easy to do an OK std with two-pass using this (https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Two-pass_algorithm) or to implement Welford's online algorithm into `torch_runstats` maybe (https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm)?

Or can just return `NaN` for the std if you only ever use the mean for now.

**Comment by Linux-cpp-lisp** [comment]
Hi @peastman ,

I've just brought this branch up-to-date with `develop` and would be very happy to get it merged for the next. I've left a few comments.

> rework the main statistics routine to support streaming calculations so it can work with any dataset class.

For the simple subset of statistics implemented here, I think this can be done without _too_ much work. The only problem is that it would call `.get()` repeatedly to stream over `AtomicData` objects so the overhead might be high with your neighborlist calculation.

I think the best solution for that is to go to the root of the problem: if you want to implement an alternative neighborlist backend (torch-nl or matscipy?) that could solve it.

File: nequip/data/_dataset/_hdf5_dataset.py
Comment (Linux-cpp-lisp) [inline nequip/data/_dataset/_hdf5_dataset.py:14]: Could you please add a docstring describing how this is meant to be used?

File: nequip/data/_dataset/_hdf5_dataset.py
Comment (peastman) [reply nequip/data/_dataset/_hdf5_dataset.py:49]: Possibly it isn't needed and that just got carried over.  I gather `AtomicDataDict.FORCE_KEY` corresponds to the actual force (the negative gradient of the energy), and not to the gradient?

File: nequip/data/_dataset/_hdf5_dataset.py
Comment (peastman) [reply nequip/data/_dataset/_hdf5_dataset.py:46]: Sounds good, I'll make the change.

**Comment by peastman** [comment]
Thanks!  I'll try to get the changes in shortly.

File: nequip/data/_dataset/_hdf5_dataset.py
Comment (Linux-cpp-lisp) [reply nequip/data/_dataset/_hdf5_dataset.py:49]: Yes exactly, `AtomicDataDict.FORCE_KEY` is the force (neg grad) and not the gradient.

**Comment by Linux-cpp-lisp** [comment]
Great, thanks @peastman !

File: nequip/data/_dataset/_hdf5_dataset.py
Comment (peastman) [reply nequip/data/_dataset/_hdf5_dataset.py:94]: It needs to be modifiable.  It's just a local variable.  The result that gets returned is a tuple.

**Comment by peastman** [comment]
After the updates, I now get an exception when I try to train a model with a HDF5 dataset.  Here's the output.

```
Number of weights: 254968
Number of trainable weights: 254968
! Starting training ...
/home/conda/feedstock_root/build_artifacts/pytorch-recipe_1660070785140/work/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:365: operator(): block: [0,0,0], thread: [72,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/home/conda/feedstock_root/build_artifacts/pytorch-recipe_1660070785140/work/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:365: operator(): block: [0,0,0], thread: [80,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
```

and then after many repetitions of similar lines,

```
Traceback (most recent call last):
  File "/home/peastman/miniconda3/envs/torchmd-net/bin/nequip-train", line 33, in <module>
    sys.exit(load_entry_point('nequip', 'console_scripts', 'nequip-train')())
  File "/home/peastman/workspace/nequip/nequip/scripts/train.py", line 102, in main
    trainer.train()
  File "/home/peastman/workspace/nequip/nequip/train/trainer.py", line 779, in train
    self.epoch_step()
  File "/home/peastman/workspace/nequip/nequip/train/trainer.py", line 912, in epoch_step
    self.batch_step(
  File "/home/peastman/workspace/nequip/nequip/train/trainer.py", line 809, in batch_step
    out = self.model(data_for_loss)
  File "/home/peastman/miniconda3/envs/torchmd-net/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/peastman/workspace/nequip/nequip/nn/_graph_model.py", line 112, in forward
    data = self.model(new_data)
  File "/home/peastman/miniconda3/envs/torchmd-net/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/peastman/workspace/nequip/nequip/nn/_rescale.py", line 144, in forward
    data = self.model(data)
  File "/home/peastman/miniconda3/envs/torchmd-net/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/peastman/workspace/nequip/nequip/nn/_grad_output.py", line 310, in forward
    data = self.func(data)
  File "/home/peastman/miniconda3/envs/torchmd-net/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/peastman/workspace/nequip/nequip/nn/_graph_mixin.py", line 356, in forward
    input = module(input)
  File "/home/peastman/miniconda3/envs/torchmd-net/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/peastman/workspace/nequip/nequip/nn/embedding/_one_hot.py", line 41, in forward
    one_hot = torch.nn.functional.one_hot(
RuntimeError: CUDA error: device-side assert triggered
```

Any idea what would cause that?

File: nequip/data/_dataset/_hdf5_dataset.py
Comment (Linux-cpp-lisp) [reply nequip/data/_dataset/_hdf5_dataset.py:94]: Ah good point, sorry--- I missed that

**Comment by Linux-cpp-lisp** [comment]
Yeah we've seen this before when the there are atom indexes in the `edge_index` that are larger than the number of atoms.

**Comment by peastman** [comment]
Where would the invalid indices come from?  Nothing in this class sets them.  The AtomicData objects get created like this:

```python
        args = {
            "pos": data[1][i],
            "r_max": self.r_max,
            AtomicDataDict.ATOM_TYPE_KEY: data[0][i],
            AtomicDataDict.TOTAL_ENERGY_KEY: data[2][i],
        }
        if self.has_forces:
            args[AtomicDataDict.FORCE_KEY] = data[3][i]
        return AtomicData.from_points(**args)
```

**Comment by peastman** [comment]
I made most of the changes, and also added unit tests.  The one thing I haven't done is switch `statistics()` to use pytorch_runstats, since it doesn't support standard deviation.  If that feature is added to it, it should be easy to make `statistics()` use it.  In the mean time, all the tests in TestStatistics are now being run on it to verify the results are correct.

**Comment by peastman** [comment]
The tests are failing because h5py isn't installed in the test environment.  I can fix it in either of two ways.  1) Add h5py as a dependency in setup.py so it will always get installed, or 2) make it skip those tests if it isn't installed.  Which do you prefer?  I suggest 1, but I don't want to add a dependency without checking first.

**Comment by Linux-cpp-lisp** [comment]
>  The one thing I haven't done is switch statistics() to use pytorch_runstats, since it doesn't support standard deviation. If that feature is added to it, it should be easy to make statistics() use it.

Online one-pass stddev might be tricky, but I can look later at how bad it would be to implement Welford's algorithm in the simple case.

> The tests are failing because h5py isn't installed in the test environment. I can fix it in either of two ways. 1) Add h5py as a dependency in setup.py so it will always get installed, or 2) make it skip those tests if it isn't installed. Which do you prefer? I suggest 1, but I don't want to add a dependency without checking first.

I don't want to add it as a mandatory dependency, so why don't we `skipif` the tests on h5py but _also_ just add a `pip install h5py` to the test workflow YAMLs?

**Comment by peastman** [comment]
It already uses Welford's algorithm.  It's mainly a matter of upstreaming the code to pytorch_runstats, though that will also add some complexity since pytorch_runstats supports features that aren't needed here.

> why don't we skipif the tests on h5py but also just add a pip install h5py to the test workflow YAMLs?

Sounds good.

**Comment by Linux-cpp-lisp** [comment]
> It already uses Welford's algorithm. It's mainly a matter of upstreaming the code to pytorch_runstats, though that will also add some complexity since pytorch_runstats supports features that aren't needed here.

Ah I see--- you mean `accumulate_by` and `reduce_dims`? Reduce dims should be easy and we can just not support `accumulate_by` for std for now, just throw `NotImplementedError`, I think that's fine.

**Comment by peastman** [comment]
This should be ready to merge, unless you need other changes first.

**Review by Linux-cpp-lisp — APPROVED** [review]

**Comment by Linux-cpp-lisp** [comment]
Looks good to me, no need to wait on statistics stuff to do a first merge--- I'll get this merged later today.  Thanks @peastman !

**Comment by Linux-cpp-lisp** [comment]
Merged!   If I manage to get around to implementing Welford's into runstats, we can revisit migrating to that and making this a "first class"  dataset in terms of feature support in another PR later. 

Thanks for your efforts on this @peastman !

**Comment by peastman** [comment]
Thank you!

**Pull request merged**
